test/karma/test-app/slot-reorder/karma.spec.ts

```diff
@@ -72,7 +75,7 @@ describe('slot-reorder', () => {
       expect(r.children[0].hasAttribute('hidden')).toBe(false);
       expect(r.children[0].getAttribute('name')).toBe('slot-b');
       expect(r.children[1].textContent.trim()).toBe('fallback default');
-      expect(r.children[1].hasAttribute('hidden')).toBe(false);
+      expect(r.children[1].hasAttribute('hidden')).toBe(true);
```

Comment:
I'm not sure I fully understand this - the implication of this code change to me is that the `'hidden'` attribute's presence/value would be changing from today's existing behavior for scoped components - can you help me understand if this is the case / why we need to change this value?

Reply:
I believe what's going on is that previously this element was being destroyed and re-created so that it didn't have the hidden attribute anymore, now it is preserved across re-renders so although the _value_ of `el.getAttribute('hidden')` is I believe consistent and what it's supposed to be we see a difference in whether or not the attribute has been set on the element. Does that make sense?

Reply:
I'm not sure, I'm still not totally clear as to how this is a result of automatic key insertion. Who/what was doing the recreation before, and who's doing it now? I suppose what I'm trying to understand here is any differences in our render cycle that might be a result of this change, and this question is just a small part of that.

Reply:
I realized that this is another situation where I think we want to bail before adding keys to the JSX. This is the component that's being tested in this file:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/test/karma/test-app/slot-reorder/cmp.tsx#L1-L40

In particular, note the `render()` method has two returns, each of which wrap some slot stuff with `<div>` tags, like so:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered">
          { ... }
        </div>
      );
    }

    return (
      <div>
        { ... }
      </div>
    );
  }
}
```

in this case we can't safely add `key` attrs to those divs, because doing so will cause the runtime to not match them up across renders. In particular, our implementation of `isSameVnode` assumes that _if keys are set_ on two vnodes then those are authoritative:

https://github.com/ionic-team/stencil/blob/148e8cd45d96e9a331a3aac307e49a2b18fef0e0/src/runtime/vdom/vdom-render.ts#L586-L604

The transformed version of the component in question here would look something like this, with keys inserted:

```tsx
export class SlotReorder {
  @Prop() reordered = false;

  render() {
    if (this.reordered) {
      return (
        <div class="reordered" key="asdfasdf1234">
          { ... }
        </div>
      );
    }

    return (
      <div key="hjklhjkl5678">
        { ... }
      </div>
    );
  }
}
```

because the keys set on those syntax nodes are _not_ equal, when the vdom assesses whether or not these nodes are equal it will see that they _have_ `key` attrs and, in that case, it will say "ok, are these equal? no? then these are not the same vnode". This will cause it to then re-render the whole thing, leading to some undesirable behavior where the `<div>` actually _should_ be kept around between re-renders so that its children can be diffed as well, rather than always being re-rendered on each render cycle.

I thought about this a bit, and I think that there's just not really a safe way that we can add keys in a case like this where there are multiple returns from the component's `render()` method. Or, at the least, I don't think we could safely do so without adding some more fancy static analysis that I dont _think_ we want to take on. Like adding keys would be safe here I think:

```tsx
export class ImageOrText {
  @Prop() showImage = false;

  render() {
    if (this.showImage) {
      return (
        <img src="..." />
      );
    }

    return (
      <span>text instead!</span>
    );
  }
}
```

because the two tags are different the vdom would already have to re-render the elements when `this.showImage` changes, so we wouldn't run into problems adding keys to the `img` and `span` tags in the JSX. But! I don't think we probably want to bite off trying to figure out when it's safe to do this. So instead I think we need to not add keys when:

1. there are two returns in a component's `render` method and
2. a component's `render` method returns a ternary

Does that all make sense? I've added some further constraints to when we transform JSX nodes and it makes this test in pass without the changes I have in this file right now.